cql3/restrictions/statement_restrictions.cc

```diff
@@ -569,17 +562,514 @@ dht::partition_range_vector statement_restrictions::get_partition_key_ranges(con
+                ck_ranges.push_back(
+                        reverse_if_reqd(
+                                last_range->transform([] (const bytes& val) { return clustering_key_prefix({val}); }),
+                                *schema->clustering_column_at(i).type));
+            } else {
+                // Prior clustering columns are equality-restricted (either via = or IN), producing one or more
+                // prefix_bounds elements.  Now we will turn each such element into a CK range dictated by those
+                // equalities and this inequality represented by last_range.  Each CK range's upper/lower bound is
+                // formed by extending the prefix_bounds element with the corresponding last_range bound, if it
+                // exists; if it doesn't, the CK range bound is just the prefix_bounds element, inclusive.
+                //
+                // For example, the expression `c1=1 AND c2=2 AND c3>3` makes lower CK bound (1,2,3) exclusive and
+                // upper CK bound (1,2) inclusive.
+                ck_ranges.reserve(prefix_bounds.size());
+                const auto extra_lb = last_range->start(), extra_ub = last_range->end();
+                for (auto& b : prefix_bounds) {
+                    auto new_lb = b, new_ub = b;
+                    if (extra_lb) {
+                        new_lb.push_back(extra_lb->value());
+                    }
+                    if (extra_ub) {
+                        new_ub.push_back(extra_ub->value());
+                    }
+                    query::clustering_range::bound new_start(new_lb, extra_lb ? extra_lb->is_inclusive() : inclusive);
+                    query::clustering_range::bound new_end  (new_ub, extra_ub ? extra_ub->is_inclusive() : inclusive);
+                    ck_ranges.push_back(reverse_if_reqd({new_start, new_end}, *schema->clustering_column_at(i).type));
+                }
+            }
+            return ck_ranges;
+        }
+    }
+    // All prefix columns are restricted by EQ or IN.  The resulting CK ranges are just singular ranges of corresponding
+    // prefix_bounds.
+    std::vector<query::clustering_range> ck_ranges(prefix_bounds.size());
+    std::transform(prefix_bounds.cbegin(), prefix_bounds.cend(), ck_ranges.begin(),
+                   std::bind_front(query::clustering_range::make_singular));
+    return ck_ranges;
+}
+
+using opt_bound = std::optional<query::clustering_range::bound>;
+
+/// Makes a partial bound out of whole_bound's prefix.  If the partial bound is strictly shorter than the whole, it is
+/// exclusive.  Otherwise, it matches the whole_bound's inclusivity.
+opt_bound make_prefix_bound(
+        size_t prefix_len, const std::vector<bytes>& whole_bound, bool whole_bound_is_inclusive) {
+    if (whole_bound.empty()) {
+        return {};
+    }
+    // Couldn't get std::ranges::subrange(whole_bound, prefix_len) to compile :(
```

Comment:
Yes.

Reply:
`whole_bound` is a vector of vectors, see above.